runtime/src/cache_hash_data.rs

```diff
@@ -43,26 +43,10 @@ impl CacheHashDataFile {
 
     /// get '&T' from cache file [ix]
     fn get<T: Sized>(&self, ix: u64) -> &T {
-        // get cache file[ix..]
-        let slice = self.get_slice::<T>(ix);
-        // return [0]
-        &slice[0]
-    }
-
-    /// get '&[T]' from cache file [ix..]
-    fn get_slice<T: Sized>(&self, ix: u64) -> &[T] {
-        let start = (ix * self.cell_size) as usize + std::mem::size_of::<Header>();
-        let item_slice: &[u8] = &self.mmap[start..];
-        let remaining_elements = item_slice.len() / std::mem::size_of::<T>();
-        assert!(
-            remaining_elements > 0,
-            "ix: {ix}, remaining_elements: {remaining_elements}, capacity: {}",
-            self.mmap.len()
-        );
-
+        let item_slice = self.get_slice_internal::<T>(ix);
         unsafe {
             let item = item_slice.as_ptr() as *const T;
-            std::slice::from_raw_parts(item, remaining_elements)
+            &*item
```

Comment:
Looks like old behaviour is UB too: `get_slice_internal` is fine since it returns `&[u8]` which is guaranteed to have 1-byte alignment. `&*item` is UB tho unless `item_slice.as_ptr()` is guaranteed to be aligned to `align_of::<T>()`.

To make this not UB you can:

- change the signature to return `Option<&T>` and return `None` if not aligned
- panic if not aligned
- return `Cow<T>` and return a copy if not aligned
- ensure that things are aligned on disk by applying padding as necessary, then loading the mmap at an aligned address so that the content stays aligned. After that, make one `get_X()` accessor for every distinct type `X` you're going to read, and make it `debug_assert!` that it's aligned